Add initial support for rjsf in pipeline properties

marthacryan opened the pull request:
Corresponding to elyra-ai/pipeline-editor#197. 

Fixes #2759 

Todo:
- [x] basic pipeline properties support
- [x] basic generic node properties support
- [x] basic "canvas properties" support
- [x] finish airflow support
- [x] link in custom components
- [x] link in validation
- [x] add in required fields
- [x] add inputpath as oneof
- [x] frontend: pipeline migration
  - [x] backend: double check that relevant test pipelines are migrated properly (once migration is finalized)
- [ ] go through code to make sure all utilities and edits etc. are all corresponding to the updated schema


Testing Follow Up TODOs:
- [x] section headers: description 
- [x] backend validation requires updates (pending some frontend changes)
- [x] pipeline JSON occasionally gets corrupted (e.g. one node's properties/values find their way into another node's `component_parameters` stanza) when connected 2 nodes (no longer seeing; uncheck if not the case)
- [x] frontend: properties don't always save if you navigate away or sometimes disappear after connecting 2 nodes - possibly related to above task
- [x] frontend: schema shows up in pipeline JSON
- [x] frontend: headers appear in pipeline properties `app_data.properties.pipeline_defaults` stanza
- [x] frontend: headers appear in node properties `component_parameters` stanzas
  - [x] backend: code can be updated (`validation.py:493-500`) after this change is made
- [x] frontend: properties not listed as required in the schema show as required in properties panel
- [x] frontend: change look of some items (see [comment](https://github.com/elyra-ai/elyra/pull/2780#issuecomment-1226803610))
- [x] frontend: pipeline editor canvas goes blank occasionally when editing array values (haven't been able to reproduce recently)
- [x] `inputpath` structure in pipeline JSON needs fixed
- [x] drag and drop of notebooks/scripts
- [x] pipeline properties - placeholders for list-valued properties
- ~[ ] user-friendly key-value inputs (env variables etc all)~ - pushing to next release
- [x] pipeline default property propagation (including display in node properties)
- [x] pipeline defaults: section headers (node defaults, generic node defaults)
- [x] custom components: component description missing
- [x] generic components: [node] label description missing
- [x] certain string property validation (frontend) erroneously require objects after connecting to another node
- [x] properties with empty strings as defaults erroneously pass frontend validation when first dragging component into pipeline
- [x] `component_source` parameter should be moved out of `component_parameters` stanza of pipeline JSON (some frontend work required in order to still display the property in properties panel)
- [x] frontend validation fails for node runtime image even if a pipeline default is supplied 
- [x] frontend: properties keep the value when changing input type (e.g. when changing from raw value to file input type, the raw value default is still shown in the value field)
- [x] frontend: airflow component tool tips not displaying all properties


Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

akchinSTC commented:
@ajbozarth - Do we think this will make the 3.11 milestone this friday?

ptitzler commented:
Not specific to RJSF migration but since the PR already modifies `canvas_properties_template.jinja2`, can we add descriptions for the [`inputs`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L22) and [`outputs`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L34) headers like we already have for [`additional properties`](https://github.com/elyra-ai/elyra/blob/4da8264406871c71aa08c4d416a206fc9299361e/elyra/templates/components/canvas_properties_template.jinja2#L49)?

![image](https://user-images.githubusercontent.com/13068832/185373318-127b2edb-2f0a-46d6-b565-0117faaedec5.png)

akchinSTC commented:
When using the papermill operator, the area for the `notebook` parameter in the component properties side panel isnot rendering, but is a required field. 
![image](https://user-images.githubusercontent.com/13340013/185502403-6203960b-2c21-46b3-9574-ec7939214906.png)

kiersten-stokes commented:
> When using the papermill operator, the area for the `notebook` parameter in the component properties side panel isnot rendering, but is a required field. 

I think I found a solution to this! Should be working now

ptitzler commented:
Re-assigned to 3.12 milestone since the enhancements that users would benefit from (e.g. https://github.com/elyra-ai/elyra/issues/2738, https://github.com/elyra-ai/elyra/issues/2750) won't be ready before cut-off according to https://github.com/elyra-ai/elyra/issues/2738#issuecomment-1219676250

ptitzler commented:
A few things I didn't see listed in the todo list in the official list.

### Pipeline properties

 - Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
 - Sections are not clearly separated (compare this to the layout used for node properties (shown below), which is somewhat better due to indentation). 
<img width="381" alt="image" src="https://user-images.githubusercontent.com/13068832/186582240-1777ea70-0095-4a4f-8d0b-bf321049cad7.png">

   Layout for node properties
   <img width="374" alt="image" src="https://user-images.githubusercontent.com/13068832/186582368-65cde0f2-64cc-436a-9064-1978e869d65f.png">

Since descriptions (for headings/properties) are now only displayed as tool tips, they are much harder to read if they are longer than a few words. In many cases a new user will likely require additional information, which we make available in the documentation. How can we make links available? Could the (?) icons perhaps serve two purposes: as a visual indication that there is info available (via tool tip) and as an HREF to a URL that would open up in a new browser window if one clicks on the icon.

ajbozarth commented:
@kiersten-stokes I pushed the fix to migration.

I also updated some of the snapshot files, one of them will need to be regenerated again after we fix the issue with persisting the schema.

As for the complex pipeline test, that one uses accessors that no longer exist to fill out the form, I tried to update it to use the new accessors, but there are none in the new form. @marthacryan

marthacryan commented:
> * Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
Fixed

> * Sections are not clearly separated (compare this to the layout used for node properties (shown below), which is somewhat better due to indentation).
Fixed

> Since descriptions (for headings/properties) are now only displayed as tool tips, they are much harder to read if they are longer than a few words. In many cases a new user will likely require additional information, which we make available in the documentation. How can we make links available? Could the (?) icons perhaps serve two purposes: as a visual indication that there is info available (via tool tip) and as an HREF to a URL that would open up in a new browser window if one clicks on the icon.
I can add this as a todo, I'd need to discuss this a little further

ptitzler commented:
> > * Heading font size is smaller than text font size (see "Pipeline Runtime" vs "Kubeflow Pipelines")
> 
> Fixed

Heading font size is still 12px vs text font size of 16px

![image](https://user-images.githubusercontent.com/13068832/188808130-e1070dcb-1a1f-4168-b882-54f395de58ba.png)

---

[generic node] adjacent buttons should not touch

![image](https://user-images.githubusercontent.com/13068832/188808968-d109d1f7-d989-49e6-9df9-7bf59289c55e.png)

---

tool tips for custom components aren't displayed properly

![image](https://user-images.githubusercontent.com/13068832/188809851-278a2ce1-5b05-4bec-93c2-af6571df4155.png)


---

combination of several issues if one doesn't enter a text value in list-valued input:
 - set this as pipeline default
    
   ![image](https://user-images.githubusercontent.com/13068832/188810812-84e545a4-8f18-4475-beaa-ae39224e45f5.png)

 - non-existent value rendered  in node properties as pipeline default
    ![image](https://user-images.githubusercontent.com/13068832/188811177-cc3cf453-9034-4a07-91a2-47b78c69ce1d.png)

  - stored as `null` in pipeline file
    
    ![image](https://user-images.githubusercontent.com/13068832/188811466-b764c7d0-d0c3-4bfd-8c99-0f67904963ff.png)

ptitzler commented:
@kiersten-stokes is this ready for testing or still in wip? https://github.com/elyra-ai/elyra/pull/2767 indicates that this PR will take care of it

![image](https://user-images.githubusercontent.com/13068832/188814000-363f2004-5b44-4f52-a986-7a88e7eb1403.png)

kiersten-stokes commented:
> @kiersten-stokes is this ready for testing or still in wip? #2767 indicates that this PR will take care of it

Yes, this is ready to test! I saw your comment on #2767 as well and added the relevant logic to the `describe` command.

marthacryan commented:
> Heading font size is still 12px vs text font size of 16px

I think I hadn't updated the link with that fix - it looks right on my end but let me know if you're still seeing it on a more recent build. 

> [generic node] adjacent buttons should not touch

Fixed

> tool tips for custom components aren't displayed properly

Fixed

> * non-existent value rendered  in node properties as pipeline default
>   ![image](https://user-images.githubusercontent.com/13068832/188811177-cc3cf453-9034-4a07-91a2-47b78c69ce1d.png)

Fixed but I need to go in and prevent empty array values from being saved to the file.

ptitzler commented:
`make install` in a clean environment fails due to linting issues:

```
yarn eslint --max-warnings=0
yarn run v1.22.4
$ eslint . --fix --ignore-path .gitignore --ext .ts,.tsx,.js --max-warnings=0

/Users/patti/elyra_dev2/martha-rjsf/elyra/packages/pipeline-editor/src/PipelineEditorWidget.tsx
  306:28  warning  Missing return type on function                                                                                        @typescript-eslint/explicit-function-return-type
  338:6   warning  React Hook useCallback has a missing dependency: 'removeNullValues'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

✖ 2 problems (0 errors, 2 warnings)

ESLint found too many warnings (maximum: 0).
error Command failed with exit code 1.
```

ptitzler commented:
@kiersten-stokes: overall file input looks good but I managed to break this scenario:
 - choose an input property that has a default value defined (e.g. for bashoperator use `output_encoding`, which is set to `utf-8`)
 - change the selection from `... a string value` to `... select a file`
 - the default value is now interpreted as a file name, which it isn't

ptitzler commented:
> > tool tips for custom components aren't displayed properly
> 
> Fixed

@marthacryan now no properties are displayed
![image](https://user-images.githubusercontent.com/13068832/189346849-1c8d91f5-731e-4f9e-9827-b03e1a5dffdf.png)

main branch output:

![image](https://user-images.githubusercontent.com/13068832/189346904-b0c2aaa1-c52a-4ed3-9a77-07135b27003c.png)

ptitzler commented:
It appears that we are now persisting a property named `component_source_header` in the `.pipeline` file. This property is only used as a visualization aid ("section header") in the UI and shouldn't be persisted:

From https://github.com/marthacryan/elyra/blob/pipeline-rjsf/elyra/templates/components/canvas_properties_template.jinja2#L123-L129
```    
"component_source_header": {
      "type": "null",
      "title": "Component Source",
      "uihints": {
        "ui:field": "header"
      }
    },
```

ptitzler commented on elyra/tests/pipeline/kfp/test_processor_kfp.py:
```diff
@@ -25,7 +25,7 @@
 from elyra.pipeline.catalog_connector import FilesystemComponentCatalogConnector
 from elyra.pipeline.catalog_connector import UrlComponentCatalogConnector
-from elyra.pipeline.component import Component
+from elyra.pipeline.component import Component, ComponentParameter
```
```suggestion
from elyra.pipeline.component import Component
from elyra.pipeline.component import ComponentParameter
```

kiersten-stokes commented:
> choose an input property that has a default value defined (e.g. for bashoperator use output_encoding, which is set to utf-8)
change the selection from ... a string value to ... select a file
the default value is now interpreted as a file name, which it isn't

@ptitzler yeah, just checked and I see this too. I think the fix will be in the frontend, but I'll double check at the meeting this AM!

marthacryan commented:
@ptitzler Thanks for checking it out again! I fixed the lint and the extra null value. I'm working on the airflow components tooltips issue

ptitzler commented:
@ajbozarth migration of the [noaa tutorial pipeline](https://github.com/elyra-ai/examples/tree/main/pipelines/dax_noaa_weather_data) fails:

![image](https://user-images.githubusercontent.com/13068832/189652106-e49f1395-32f7-4d51-8b82-ed70641d0d1f.png)

ptitzler commented:
The validation message is invalid if one does not specify a file as string input:

![image](https://user-images.githubusercontent.com/13068832/189844417-bb7f55a6-b80f-4393-a08b-8fc1cca20650.png)

Note the value of `value`, which is not what the user provided as input. The expected value would be null (undefined).

ptitzler commented:
Tooltip for custom nodes in Airflow are now looking good ...
![image](https://user-images.githubusercontent.com/13068832/189847917-23a378e2-3dad-4c4f-9d8a-6d46f4427c0f.png)

For property values that are provided via input files (`URL` in node `download file` in the example below), can we prepend the file name with a string that indicates that we are displaying a reference, not the value? The `count rows` node is an example of a property that is also defined via reference, specifically a reference to the output of an upstream node:

![2022-09-13_10-16-26 (1)](https://user-images.githubusercontent.com/13068832/189849100-512f163e-a910-4801-8a7c-787bdcf29964.gif)

kiersten-stokes commented:
> The validation message is invalid if one does not specify a file as string input:

Good point! I've fixed validation to allow an empty string for file input types for non-required properties. From here, the processor behaves the same way as an empty string for raw input as well.

ptitzler commented:
> > The validation message is invalid if one does not specify a file as string input:
> 
> Good point! I've fixed validation to allow an empty string for file input types for non-required properties. From here, the processor behaves the same way as an empty string for raw input as well.

Thank you! Looking good. It appears to preserve default values, which I would expect to happen since the user didn't provide a property value.

kevin-bates commented:
Congratulations (and thank you :heart:) to everyone that worked on this - it's one for the ages! :tada:

thesuperzapper commented:
I would also love to thank everyone for their amazing effort!!!

❤️
